File: py/keepintouch/schedule.py
import datetime
import random
import math

from keepintouch import util

class Scheduler(object):
    def schedule(self, entries):
        pass
        
class BacklogScheduler(Scheduler):
    @classmethod
    def _appraise(cls, entry):
        return entry.last_contacted + datetime.timedelta(days=entry.interval)

    @classmethod
    def schedule(cls, entries):
        today = util.todays_date()

        appraised_entries = []
        for entry in entries:
            due = cls._appraise(entry)
            if due < today:
                appraised_entries.append((entry, due))

        appraised_entries.sort(key=lambda p: p[1])

        return [p[0] for p in appraised_entries]

class WeightScheduler(Scheduler):
    DEFAULT_WEIGHT = 0.75

    def __init__(self, weight=DEFAULT_WEIGHT):
        self.weight = weight

    def _appraise(self, entry):
        return math.floor(
            self.weight *                     # A portion of...
            entry.interval +                  #   ...the interval and
            (1 - self.weight) * 2 *           # a portion of...
            random.random() * entry.interval  #   ...random above or below original
            )

    def schedule(self, entries):
        return sorted(entries, key=self._appraise)

class RandomScheduler(Scheduler):
    @classmethod
    def schedule(cls, entries):
        shuffled_entries = [e for e in entries]

        random.shuffle(shuffled_entries)

        return shuffled_entries

File: hs/KeepInTouch/Schedule.hs
module KeepInTouch.Schedule
(
  --
  Weight(..)
, defaultWeight

, Backlog(..)

, Shuffle(..)
) where

import Data.Function(on)
import Data.List(sort)
import Data.Time.Calendar(Day,addDays)
import System.Random(randoms,RandomGen)

import KeepInTouch.Type(Entry(..),Scheduler(..))
import KeepInTouch.Util(shuffle)

data Weight a = Weight
    { weight    :: Float
    , generator :: a
    }

-- Default weight 
defaultWeight :: Float
defaultWeight = 0.25

-- Entries sorted by lastContacted + a portion of the interval where the portion
--   of the interval will be randomly in the range:
--
--   interval -/+ (weight * interval)
--
instance (RandomGen a) => Scheduler (Weight a) where
    schedule (Weight weight generator) =
      let
        -- Pair with random values
        percentages = zip (randoms generator)

        -- lastContacted + (1 - weight) * interval +/- weight * random
        weightedValue :: (Float, Entry) -> (Day, Entry)
        weightedValue (randomPercent, entry) =
          let
            intervalF = fromIntegral $ interval entry

            normalWeighted = intervalF * (1 - weight)
            randomWeighted = intervalF * weight * 2 * randomPercent

            intervalDays = round $ normalWeighted + randomWeighted
            startDay = lastContacted entry
          in
            (addDays intervalDays startDay, entry)
      in
        -- lastContacted with random percentage of interval
        map snd . sort . map weightedValue . percentages

data Backlog = Backlog
    { today :: Day
    }

instance Scheduler Backlog where
    -- Entries due (lastContacted + interval) before today
    schedule (Backlog thisDay) =
      let
        overdue e = addDays (interval e) (lastContacted e)
        sortedWithOverdue = sort . map (\e -> (overdue e, e))

        beforeToday = fst . break ((thisDay <=) . fst)

        toEntries = map snd
      in
        toEntries . beforeToday . sortedWithOverdue

data Shuffle a = Shuffle a

instance (RandomGen a) => Scheduler (Shuffle a) where
    schedule (Shuffle gen) = shuffle gen
